articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,313 @@
+On systems where the implementation is handled for us by DDS (like with Connext), we use `rS_` as the common prefix.
+
+Here is a non-exhaustive list of prefixes:
+
+| ROS Subsystem        | Prefix |
+|----------------------|--------|
+| ROS Topics           | rt_    |
+| ROS Service Request  | rq_    |
+| ROS Service Response | rs_    |
+| ROS Service          | rS_    |
+| ROS Parameter        | rp_    |
+| ROS Action           | ra_    |
+
+### Substitution of the Namespace Delimiter
+
+The namespace delimiter in ROS 2 topic and service names, a forward slash (`/`), will be replaced with double underscores (`__`).
+The leading forward slash (`/`) is not replaced with double underscores (`__`), and is replaced instead by the ROS specific name prefix.
+
+### ROS to DDS Name Conversion Examples
+
+Here are some examples of translations between ROS topic names and DDS topic names:
+
+| ROS Name          | Subsystem | DDS Name              |
+|-------------------|-----------|-----------------------|
+| `/foo`            | Topic     | `rt_foo`              |
+| `/foo/bar`        | Topic     | `rt_foo__bar`         |
+| `/_foo/bar`       | Topic     | `rt__foo__bar`        |
+| `/foo/_bar`       | Topic     | `rt_foo___bar`        |
+| `/_foo/_bar`      | Topic     | `rt__foo___bar`       |
+| `/_foo/_bar/_baz` | Topic     | `rt__foo___bar___baz` |
+| `/foo/_/bar`      | Topic     | `rt_foo_____bar`      |
+
+## Concerns/Alternatives
+
+This section lists concerns about the proposed design and alternatives that were considered.
+
+### Alternative ROS to DDS Mapping
+
+There was some discussion of alternatives and concerns with respect to the ROS -> DDS translation.
+
+#### Capital Letter Substitution Alternative
+
+Since the forward slash (`/`) is replaced with double underscores (`__`) when translating to DDS from ROS topic names, two characters out of the 256 character limit are lost with each additional namespace.
+One proposed alternative was to add a constraint that ROS topic names could not use capital letters, and then capital letters could be used as the stand in for the forward slashes (`/`).
+
+Trade-offs:
+
+- Uses one less character per namespace and makes it easier to calculate the maximum length of a ROS topic or service name.
+- Prevents users from using capital letters in their names, which is constraining and might be a problem for backwards compatibility with ROS 1 topics and services.
+
+Rational:
+
+Preventing users from using capital letters was too constraining for the added benefit.
+
+#### Prefix with Double Underscores
+
+This alternative differs only in that it uses two underscores in the prefix, i.e. `rt__` rather than `rt_`.
+
+Trade-offs:
+
+- More consistent with replacement of other forward slashes (`/`)
+- Uses one more character
+
+Rational:
+
+Slight preference given to the shorter alternative.
```

Comment:
> I'm not sure I agree because I can't imagine why we'd need more characters (but it's hard to prove a negative). Especially given that we can start using the second character as an op-code if we need more options and then we have around 30 unique values (a-z + 0-9 - the ones we've already reserved). That would be a hit to readability and obviousness, but it's also something I don't think we'll have to do.
> 
> Also, when considering reserving more characters, I think that we should assume that we'll never give the characters back, because I think it would be a painful process to do so. So much so that it would never be worth rolling out. It would basically be a wire level breaking change to change the prefix. I'm not arguing that we shouldn't reserve space because of this, but it's like a mail-in rebate to me: I just assume the price does not include the rebate and that I'm not going to get it back, but then if I do it's icing on top.

Indeed we shouldn't assume they'll come back , and we don't need to use them in the prefix. If we just enforce that we leave empty space at the end of the strings it's a forward compatible change. All old topics will still work. The value in this that to users if we allow them x elements in their names, it's a non breaking change to allow them x+3 elements in their names. It's a potentially crippling change if we change it to allow x-3 which is almost impossible to audit if there are dynamically generated topic names. 

I agree that there are no strong use cases right now for going beyond the 32 characters. But there are some things we could consider in the future such as embedded ROS version semantics into the prefix such that you could differentiate versions of ROS running over DDS in parallel etc. Leaving ourselves some extra bits might be worth it. 

> 
> If we really wanted to preserve characters we'd want to use a single separator with an escape character.
> What does this mean?

For example:
If '0' is prefixed by '_' then '0' else '/'

* a_0b -> a0b
* a0_b -> a/_b
* a_00b -> a0/b
* a__0b -> a_0b
* a_____b -> a_____b
* a0000000 -> b invald empty namespaces
* a_0_0_0_0b -> a0000b

Or looking the other way.

* trailing 0: a0/b -> a_00b
* leading underscore: a/_b -> a0_b
* zero after underscore: a_0b -> a__0b 
* zero before underscore: a0_b -> a_0_b

Overhead 1 character per namespace separator + 1 character per instance of 0 in the string. 

This is a lower overhead than 2 characters per namespace separator assuming that there are more namespace separators than zeros. But i'm not advocating for this but saying that if we need more space.

We could pick any other character such as a capital letter which is theoretically invalid in the ROS space and decrease the number of collisions. We've been talking about allowing them since we haven't enforced it before. 

> 
> The requirement that there's not trailing _ and leading numbers could make most separators only single width.
> Can you give an example? I don't see how we could use those rules to implement a separator (I assume you mean the namespace delimiter).

Zero can never start a token and underscore can never end a token thus _0 can never be at the boundary. It limits where it could be but on expanding it out, it's not limited enough to avoid the need for an escape character. I think using the capitals would be a better optimization since they should appear less. 

But again I think we should use the `__` for clarity.